cdc/split.cc

```diff
@@ -149,101 +152,152 @@ static gc_clock::duration get_ttl(const row_marker& rm) {
     return rm.is_expiring() ? rm.ttl() : gc_clock::duration(0);
 }
 
-static
-std::map<std::pair<api::timestamp_type, gc_clock::duration>, row_update>
-extract_row_updates(const row& r, column_kind ckind, const schema& schema) {
-    std::map<std::pair<api::timestamp_type, gc_clock::duration>, row_update> result;
-    r.for_each_cell([&] (column_id id, const atomic_cell_or_collection& cell) {
-        auto& cdef = schema.column_at(ckind, id);
-        if (cdef.is_atomic()) {
-            auto view = cell.as_atomic_cell(cdef);
-            auto timestamp_and_ttl = std::pair(
-                    view.timestamp(),
-                    view.is_live_and_has_ttl() ? view.ttl() : gc_clock::duration(0)
-                );
-            result[timestamp_and_ttl].atomic_entries.push_back({id, atomic_cell(*cdef.type, view)});
-            return;
+using change_key_t = std::pair<api::timestamp_type, gc_clock::duration>;
+
+template <typename V>
+struct extract_collection_visitor {
+private:
+    const column_id _id;
+    std::map<change_key_t, row_update>& _updates;
+
+    nonatomic_column_update& get_or_append_entry(api::timestamp_type ts, gc_clock::duration ttl) {
+        auto& updates = this->_updates[std::pair(ts, ttl)].nonatomic_entries;
+        if (updates.empty() || updates.back().id != _id) {
+            updates.push_back({_id});
         }
+        return updates.back();
+    }
+
+    void cell(bytes_view key, const atomic_cell_view& c) {
+        auto& entry = get_or_append_entry(c.timestamp(), get_ttl(c));
+        entry.cells.emplace_back(to_bytes(key), atomic_cell(*static_cast<V&>(*this).get_value_type(key), c));
```

Comment:
Honestly, why not either just make it virtual (compiler will eliminate it anyway), or use partial specialization? (Declare the function in the template and specialize it for the two instantiations)

Reply:
It == `get_value_type`

Reply:
Are you sure it would eliminate it? I wanted to avoid virtual calls as much as possible (wouldn't I need to allocate the visitor?)

I'm gonna try the partial specialization, I didn't think of it.

Reply:
I figured I like the CRTP way more than the alternatives.
However, since it's not obvious where the `get_value_type` method comes from at the call point, and perhaps the pattern is no so well known, I added a comment describing the situation:

```
    /* To copy a value from a collection/non-frozen UDT (in order to put it into a bucket) we need to know the value's type.
     * The method of obtaining the type depends on the collection type; in particular, for non-frozen UDT, each value
     * might have a different type, thus in general we need a method that, given a key (identifying the value in the collection),
     * returns the value' type.
     *
     * We use the `Curiously Recurring Template Pattern' to avoid performing a dynamic dispatch on the collection's type for each visited cell.
     * Instead we perform a single dynamic dispatch at the beginning, when encountering the collection column;
     * the dispatch provides us with a correct `get_value_type` method.
     * See `extract_row_visitor::collection_column` where the dispatch is done.

    data_type get_value_type(bytes_view);
    */

    void cell(bytes_view key, const atomic_cell_view& c) {
        auto& entry = get_or_append_entry(c.timestamp(), get_ttl(c));
        entry.cells.emplace_back(to_bytes(key), atomic_cell(*static_cast<V&>(*this).get_value_type(key), c));
    }
```

Reply:
I've done enough "CRTP" (hooray for acronyms - HFA) to try to avoid them if possible. But it is hardly a dealbreaker, so sure...